Command line option for adding secrets to one-shot task

MarcPer opened the pull request:
Signed-off-by: Marcelo Pereira <marcelovitor.pereira@gmail.com>

<!--- Provide a general summary of your changes in the Title above -->

## Description
The `--secret` or `-s` flag was included for adding secrets to a one-shot service. The syntax is
```
jaas --image my_image --secret secret_name=secret_file --secret
```
, where `secret_name` is the name of a secret that was previously created (an error message is returned if a secret with the given name is not found) and `secret_file` is the name of the file with the secret that will be available inside the container. All such files are available in the container under the `/run/secrets/` directory.

The flag supports multiple secrets, as follows:
```
jaas --image my_image --secret secret_name=secret_file --secret another_secret=another_secret_file
```

## Motivation and Context
This addresses the issue raised in https://github.com/alexellis/jaas/issues/14 and allows for one-shot tasks to have access to docker secrets.

## How Has This Been Tested?
This was tested on Ubuntu 18.04 with Go version go1.10.3 linux/amd64. Tests were done by running the command in the following cases:
- The command was run without the `--secret` flag
- Command was run with the `--secret` flag referencing a secret that didn't exist, raising an error
- Command was run with a `--secret` flag referencing a secret that did exist, and the `--command` flag was used to `cat` the contents of the secret file
- The command was run with multiple `--secret` flags and `cat` was used to print the contents of each of the files.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

alexellis commented:
Hi @MarcPer thank you for your interest in JaaS and for taking the time to follow the contribution guide 👍 

I've taken a quick look at the proposed PR, but I think this syntax should be only the name and not an additional file:

```
--secret secret-name1 --secret secret-name2
```

Then those secrets will appear under /run/secrets/secret-name1, /run/secrets/secret-name2 and etc. Do you have a use case where this would not be sufficient in terms of an API?

Let me know your thoughts.

Alex

alexellis commented on cmd/run.go:
```diff
@@ -171,7 +175,47 @@ func runTask(taskRequest TaskRequest) error {
+
+	spec.TaskTemplate.ContainerSpec.Secrets = []*swarm.SecretReference{}
+	for _, serviceSecret := range taskRequest.Secrets {
+		parts := strings.Split(serviceSecret, "=")
```
I think we have this code already, please can you look for a generic method or create one if there is none?

openfaas/faas-cli also has lots of similar parsing code.

MarcPer replied:
The iteration plus splitting of the argument is done elsewhere in the code. However, since I'll change the syntax to be simpler, splitting won't be necessary anymore. I'll have a look at what else is being repeated in the code and try to extract a function.

alexellis commented on cmd/run.go:
```diff
@@ -171,7 +175,47 @@ func runTask(taskRequest TaskRequest) error {
+					Name: parts[1],
+					UID:  "0",
+					GID:  "0",
+					Mode: os.FileMode(0400),
```
Why mode 0400, what does this signify?

MarcPer replied:
This is the UNIX permissions notation and means the _/run/secrets_ file inside the container can only be read by its owner, the `root` user, in this case. Perhaps I should add a comment or find a notation that is more self-explanatory. I'll have a look at the alternatives.

MarcPer commented:
Thanks for the review, @alexellis.

I think you are right; there is no need for the extended syntax. For some reason, I thought it was not possible to have a dot in a secret name, but I just tested that and it _is_ possible. I'll stick to the simpler syntax.

alexellis commented:
Thank you, looking forward to the updated PR.

MarcPer commented:
The syntax now is simpler: the argument to `--secret` is both the name of the secret and the file name inside the container.

I've changed the file permission to `0444`, so the secret can be read by any user _inside the container_. This is so `jaas` can be used with docker images that enforce a non-root user, when such a user still needs access to the secret.

I've added a comment explaining what `0444` means. We could create a constant for this, but the `os` package itself doesn't care much about making these permission bits more clear. All examples use these integers to set permissions.

About extracting the parsing into a function, I guess the extra code doesn't overlap so much with the rest anymore. But I wouldn't mind creating a PR later for adding unit tests and breaking up the `runTask` into smaller chunks of reusable code.

Finally, I've noticed that the documentation on the `--remove` flag could be improved. Removing the service is currently the default, but README.md states that for removing the service, the flag needs to be passed. Should I open another PR only for this?

alexellis commented:
> Should I open another PR only for this?

Yes please

I'll look into this PR soon.

alexellis commented on README.md:
````diff
@@ -113,6 +113,29 @@ $ export auth='{
+```
+# echo "p4ssword" > pass_file
+# docker secret create my_secret pass_file
+# jaas run --image alexellis2/href-counter:latest --env url=http://blog.alexellis.io/ --secret my_secret --command "cat /run/secrets/my_secret"
````
I guess we could just change this to `alpine:3.7` and keep the suggested command without env-vars?

alexellis replied:
```
jaas run --secret=my_secret --image alpine:3.7 --command="cat /run/secrets/my_secret"
```

alexellis commented on cmd/run.go:
```diff
@@ -171,6 +173,36 @@ func runTask(taskRequest TaskRequest) error {
 
+	secretList, err := c.SecretList(context.Background(), types.SecretListOptions{})
+
+	spec.TaskTemplate.ContainerSpec.Secrets = []*swarm.SecretReference{}
```
I know the style wasn't followed elsewhere, but do you think we can extract a function for this?

alexellis commented:
Let's get the changes in a follow-up PR?

MarcPer commented:
Sure thing.

alexellis commented:
Btw what are your thoughts on #32 ? Interesting for you?

MarcPer commented:
Hey, sorry for the delay. It does seem interesting. I see you want to decouple from Swarm - do you have other use cases in mind? For running remote docker commands, we can still use the `DOCKER_HOST` environment variable.

I think it would be nice if only the server part had access to `docker`, limiting access for the client to only what is necessary - say, for a CI server to trigger a specific one-time task. I'm thinking something like the server defines what the client can and cannot do.

Looking forward to your consideration.
